slowquant/qiskit_interface/wavefunction.py

```diff
@@ -293,19 +314,38 @@ def change_primitive(self, primitive: BaseEstimator | BaseSampler, verbose: bool
         self.QI.total_device_calls = 0
         self.QI.total_shots_used = 0
         self.QI.total_paulis_evaluated = 0
-        self.QI.cliques = Clique()
+        self.QI._reset_cliques(verbose=verbose)  # pylint: disable=protected-access
         self.QI._Minv = None  # pylint: disable=protected-access
-        self.QI._primitive = primitive  # pylint: disable=protected-access
-        # IMPORTANT: Shot number in primitive gets always overwritten if a shot number is defined in QI!
-        if self.QI.shots is not None and verbose:
-            print(
-                "Number of shots defined in new primitive are ignored as there is a number defined in the QI of ",
-                self.QI.shots,
-            )
-            print("If you want to change the number of shots, do this manually.")
-            print("Set the number of shots manually to None if you run an ideal simulator.")
-        self.QI.shots = self.QI.shots  # Redo shot check with new primitive
-        self.QI.ISA = self.QI.ISA  # Redo ISA parameter check
+
+        # Reset circuit and initiate re-transpiling
+        ISA_old = self.QI.ISA
+        self._reconstruct_circuit()  # Reconstruct circuit but keeping parameters
+        self.QI.ISA = ISA_old  # Redo ISA including transpilation if requested
+        self.QI.shots = self.QI.shots  # Redo shots parameter check
+
+        if verbose:
+            self.QI.get_info()
+
+    def change_shots(self, shots: int | None) -> None:
```

Comment:
It is not better to have this being used as `WF.QI.shots = 10000`, and then keep doing all the QI logic inside QI.

Reply:
Yes. Up to the user. This is a save option for the user to change the shots if they dont understand QI. 
E.g. they forget to reset cliques or so if they want to run with a new shot number. But one does not need to use the function.

Reply:
I mean shouldnt the clique not just reset if the number of shots is changed in QI?

My main concern with this, is that the maintainability of the code becomes a lot worse.
A change in QI, might break the methods in WF that essentially does QI logic.

Reply:
Agreed. I can put the clique reset into shots.